add a Helm post-delete-hook for Rancher

jiaqiluo opened the pull request:
Add a post-delete-hook for cleaning up the apps left after uninstalling Rancher.
As of now, it will delete 
- the `helm-operation-*` pods
- the following apps from the cluster: 
```
- fleet
- fleet-agent
- rancher-operator
- rancher-webhook
```

see this comment for the behavior when the job fails: https://github.com/rancher/rancher/pull/31670#issuecomment-800437206


issue: https://github.com/rancher/rancher/issues/30924

rawmind0 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,34 @@
+    echo "--- deleting the app [$app] (namespace: $namespace)"
+    helm uninstall "$app" -n "$namespace"
+    c=0
+    # wait 120 sections in total for an uninstallation to be finished
```
typo here

```
-    # wait 120 sections in total for an uninstallation to be finished
+    # wait 120 seconds in total for an uninstallation to be finished
```

jiaqiluo replied:
fixed.

rawmind0 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,34 @@
+      sleep 6
+      c=$((c+1))
+    done
+    echo "uninstalled" "[$app]" "in the namespace" "[$namespace]"
```
The scripts is waiting up to 120 seconds for every single app being uninstalled. I think that the exit of the `while` loop should be controlled to check if the app has been properly uninstalled or has got timeout

```
-    echo "uninstalled" "[$app]" "in the namespace" "[$namespace]"
+    if [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]]; then
+      echo "timeout uninstalling" "[$app]" "in the namespace" "[$namespace]"
+      continue
+    fi
+    echo "uninstalled" "[$app]" "in the namespace" "[$namespace]"
```

rawmind0 commented on chart/templates/post-delete-hook-configMap.yaml:
```diff
@@ -0,0 +1,14 @@
+    "helm.sh/hook-delete-policy": hook-succeeded
+data:
+  post-delete-hook.sh: |-
+{{ $.Files.Get "scripts/post-delete-hook.sh" | indent 4 }}
```
Shouldn't this template rendering be conditioned by `Values.postDelete.enabled` value??

```
+{{- if .Values.postDelete.enabled }}
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Values.postDelete.name}}-script
  namespace: {{ .Release.Namespace }}
  labels:
{{ include "rancher.labels" . | indent 4 }}
  annotations:
    "helm.sh/hook": post-delete
    "helm.sh/hook-weight": "1"
    "helm.sh/hook-delete-policy": hook-succeeded
data:
  post-delete-hook.sh: |-
{{ $.Files.Get "scripts/post-delete-hook.sh" | indent 4 }}
+{{ end }}
```

jiaqiluo replied:
Yes, you are right. The condition was added.

rawmind0 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,34 @@
+    helm uninstall "$app" -n "$namespace"
+    c=0
+    # wait 120 sections in total for an uninstallation to be finished
+    while [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]] && [[ $c -lt 20 ]]; do
```
Have you considered to set retries configurable from values?? I mean, here is hardcoded 20 retries every 6 seconds, 120 seconds. Would make sense to add a a new values var `.Values.postDelete.retries = 20` to make it configurable??

```
-    while [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]] && [[ $c -lt 20 ]]; do
+    while [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]] && [[ $c -lt $retries ]]; do
```

rawmind0 commented on chart/templates/post-delete-hook-job.yaml:
```diff
@@ -0,0 +1,42 @@
+              name: config-volume
+          env:
+            - name: namespaces
+              value: {{ .Values.postDelete.namespaceList | join " " }}
```
Adding retries values if make sense

```
+            - name: retries
+              value: {{ .Values.postDelete.retries }}
```

rawmind0 commented on chart/values.yaml:
```diff
@@ -136,3 +136,15 @@ customLogos:
+  namespaceList:
+    - fleet-system
+    - cattle-system
+    - rancher-operator-system
```
Adding retries values if make sense

```
+  # Number of retries to unsinstall apps. Post delete hook will retry every 6 seconds, 120 seconds total by default
+  retries: 20
```

rawmind0 reviewed (commented):
@jiaqiluo , my 2 cents on this PR. Please, take a look to review comments.

jiaqiluo commented:
@rawmind0  Thank you for your feedback! I have addressed all of them.

Instead of adding a variable for the count of retries, I added a variable for the total waiting time (in seconds). After the script runs the `helm uninstall` command, it repeats to check if all pods for the app are deleted every 5s until it hits the timeout.

Another thing to mention is that, in the current design, the bash script will always exit with code 0, so the job will always succeed and then delete itself. This behavior is controlled by the following annotation in the job
```
"helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
```

Do you think the script should exit with a non-zero value if it failed to uninstall any app? If the script fails, should we restart the job and keep it in the cluster if it still fail eventually?

rawmind0 commented:
> Instead of adding a variable for the count of retries, I added a variable for the total waiting time (in seconds). After the script runs the helm uninstall command, it repeats to check if all pods for the app are deleted every 5s until it hits the timeout.

No problem, it's more clear in the way you did it.  :)

> Do you think the script should exit with a non-zero value if it failed to uninstall any app? If the script fails, should we restart the job and keep it in the cluster if it still fail eventually?

I see. I think that the script should exit with a non-zero value. The uninstallation process should be run "manually" again if fails. If unsure, may be making it configurable would be a good option. It'd be just add a new bool var `postDelete.ignoreTimeoutError`, then exit non zero if false or continue if true at the script.

jiaqiluo commented:
@rawmind0 Thank you for your feedback!

Here is what I am planning to do:
- add a new bool var `postDelete.ignoreTimeoutError`. When a timeout happens, the script will exit non-zero if `postDelete.ignoreTimeoutError` is false, or continue if it is true 
- at the Job level, if `postDelete.ignoreTimeoutError` is false and the pod exits at the failed/error state, the `job` (where the pod comes from) and the `configMap` (where the script comes from) will be kept for future investigation. 

@cbron What do you think? Does that sound good to you?

cbron commented on chart/values.yaml:
```diff
@@ -136,3 +136,17 @@ customLogos:
+  name: rancher-post-delete
+  image:
+    repository: rancher/shell
+    tag: v0.1.6
```
This will not get pulled into images.txt if the primary shell gets updated. That is from here: 
https://github.com/rancher/rancher/blob/master/pkg/settings/setting.go#L94
https://github.com/rancher/rancher/blob/master/pkg/image/resolve.go#L248

Can we use the agent ? I can't think of a way to get the shell into the chart consistently. Unless we hardcode it into the requirements struct there but that feels bad.

cbron replied:
Actually now that I think about it, I'm not sure the agent image exists on the local cluster in a default (non-airgap/public) setup, do you know ? It would suck to have to pull an image to start a post-delete hook. Its likely that the shell image would have been pulled into local, and if not its 1/4th the size of the agent. Is there anything else we can guarantee is on the local cluster ?

jiaqiluo replied:
This image `rancher/shell` should already be in the local cluster because the `helm-operation` pod uses it to install Helm charts in the local cluster. 
```
> k -n cattle-system describe pod helm-operation-jp4m2 | grep -i image
    Image:         rancher/shell:v0.1.6
    Image ID:      docker.io/rancher/shell@sha256:bc884e3b37a737a7e411766c42705fbe477d6219c72ec781efd8438d8fb267d0
    Image:         rancher/shell:v0.1.6
    Image ID:      docker.io/rancher/shell@sha256:bc884e3b37a737a7e411766c42705fbe477d6219c72ec781efd8438d8fb267d0
  Normal  Pulled     6m58s  kubelet            Container image "rancher/shell:v0.1.6" already present on machine
  Normal  Pulled     6m58s  kubelet            Container image "rancher/shell:v0.1.6" already present on machine
```

Still, the problem is that how we can get the image dynamically instead of hardcoded in the value.yaml file.

jiaqiluo replied:
> This will not get pulled into images.txt if the primary shell gets updated. That is from here:
> [https://github.com/rancher/rancher/blob/master/pkg/settings/setting.go#L94](https://github.com/rancher/rancher/blob/master/pkg/settings/setting.go?rgh-link-date=2021-03-16T02%3A36%3A58Z#L94)
> [https://github.com/rancher/rancher/blob/master/pkg/image/resolve.go#L248](https://github.com/rancher/rancher/blob/master/pkg/image/resolve.go?rgh-link-date=2021-03-16T02%3A36%3A58Z#L248)
> 
> Can we use the agent? I can't think of a way to get the shell into the chart consistently. Unless we hardcode it into the requirements struct there but that feels bad.


I have 2 possible solutions so far:
1. add some logic to the file [scripts/chart/build](https://github.com/rancher/rancher/blob/master/scripts/chart/build) to get the image and tag from the following line in [pkg/settings/setting.go](https://github.com/rancher/rancher/blob/2dca0c67b48331c23c05c025208bacfc258d45bd/pkg/settings/setting.go#L94)
```
	ShellImage                        = NewSetting("shell-image", "rancher/shell:v0.1.6")
```

2. add a comment to the above line in `pkg/settings/setting.go`:  we need to update the `postDelete.image` section in the file `/chart/values.yaml` for any change on this line.

cbron replied:
1. could work but it needs to load the entire rancher codebase for it to work, so that would take good amount of time.
2. If someone forgets then airgap won't work for this, much safer to add to requirements struct and also a comment there. That seems like best solution for now.

cbron commented on chart/templates/post-delete-hook-ServiceAccount.yaml:
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.postDelete.enabled }}
```
can you lowercase the filename, uppercase looks weird.

jiaqiluo replied:
Existing files in the directory (/chart/templates) also use CamelCase, so I thought it is the pattern I should follow.  

here is the list of existing ones: 
```
clusterRoleBinding.yaml
issuer-letsEncrypt.yaml
serviceAccount.yaml
```
Should I correct them as well?

cbron replied:
Ok if its a pattern just follow those.

cbron commented on chart/templates/post-delete-hook-ServiceAccount.yaml:
```diff
@@ -0,0 +1,13 @@
+  annotations:
+    "helm.sh/hook": post-delete
+    "helm.sh/hook-weight": "1"
+    "helm.sh/hook-delete-policy": hook-succeeded
```
Interesting, I hadn't thought of using the delete policy on random objects like this. Should work.

cbron commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,38 @@
+
+set -e
+
+echo "Preparing for uninstalling apps left by Rancher in the following namespaces:" "$namespaces"
```
```suggestion
echo "Uninstalling Rancher resources in the following namespaces:" "$namespaces"
```

jiaqiluo replied:
fixed.

cbron commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,38 @@
+set -e
+
+echo "Preparing for uninstalling apps left by Rancher in the following namespaces:" "$namespaces"
+for namespace in $namespaces; do
```
Does namespaces actually get inherited from the env here ? I'm surprised its in the shells context.

jiaqiluo replied:
yes, the script gets the environment variable from the container.

cbron commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,38 @@
+      echo "waiting for pods of [$app] to be terminated ..."
+      sleep 5
+      t=$((t + 5))
+    done
```
This code looks familiar :)

cbron reviewed (commented):
Yes but I would track which of all apps failed and list them below and exit 1, not just exit 1 as soon as one is hit. Alternative may be to use k8s job spec to delete job either way after 1 hour, that was in beta and non-working by default last I checked. 
Also what is value flag default, false ? I'd say false. 
They can't really re-run the job manually, they would only use it for error logs to see which job failed.

cbron commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,38 @@
+  done
+done
+
+echo "The job is done."
```
This should list failed jobs if any, or:
```suggestion
echo "Cleanup finished successfully."
```

jiaqiluo replied:
fixed.

jiaqiluo commented:
@rawmind0 @cbron 
With setting `job.sepc.backoffLimit = 3` and `job.spec.template.spec.restartPolicy = Never`, the job will not restart failed pod, instead, it creates a new pod up to 3 times. If all three pods fail, the job will fail and keep all existing pods for users to check logs. 

Here are the results of a failed run:
(I simulate a failure by adding the `--dry-run` flag to the `helm uninstall` command and setting `timeout = 9` ) 

- helm fails to run the post-delete job (but Rancher itself is uninstalled successfully)
```
> helm uninstall rancher -n cattle-system
Error: uninstallation completed with 1 error(s): job failed: BackoffLimitExceeded
```

- the job and its pods are left in the cluster 
```
> k get job -n cattle-system
NAME                  COMPLETIONS   DURATION   AGE
rancher-post-delete   0/1           4m7s       4m7s

> k get pods -n cattle-system
NAME                              READY   STATUS    RESTARTS   AGE
rancher-post-delete-phrbc         0/1     Error     0          17m
rancher-post-delete-v9plr         0/1     Error     0          18m
rancher-post-delete-xzvqg         0/1     Error     0          17m
```
- we can check the logs for failures
```
> k -n cattle-system logs rancher-post-delete-xzvqg
Uninstalling Rancher resources in the following namespaces: fleet-system cattle-system rancher-operator-system
--- Deleting the app [fleet] in the namespace [fleet-system]
release "fleet" uninstalled
successfully uninstalled [fleet] in the namespace [fleet-system]
--- Skip the app [fleet-crd] in the namespace [fleet-system]
--- Deleting the app [rancher-webhook] in the namespace [cattle-system]
release "rancher-webhook" uninstalled
waiting 5 seconds for pods of [rancher-webhook] to be terminated ...
waiting 5 seconds for pods of [rancher-webhook] to be terminated ...
timeout uninstalling [rancher-webhook] in the namespace [cattle-system]
--- Deleting the app [rancher-operator] in the namespace [rancher-operator-system]
release "rancher-operator" uninstalled
waiting 5 seconds for pods of [rancher-operator] to be terminated ...
waiting 5 seconds for pods of [rancher-operator] to be terminated ...
timeout uninstalling [rancher-operator] in the namespace [rancher-operator-system]
--- Skip the app [rancher-operator-crd] in the namespace [rancher-operator-system]
------ Summary ------
Succeeded to uninstall the following apps: fleet
Failed to uninstall the following apps: rancher-webhook rancher-operator
```

jiaqiluo commented on pkg/settings/setting.go:
```diff
@@ -92,6 +92,7 @@ var (
 	PartnerChartDefaultBranch         = NewSetting("partner-chart-default-branch", "main")
 	FleetDefaultWorkspaceName         = NewSetting("fleet-default-workspace-name", "fleet-default") // fleetWorkspaceName to assign to clusters with none
 	ShellImage                        = NewSetting("shell-image", "rancher/shell:v0.1.6")
+	HelmPostDeletionImage             = NewSetting("helm-post-deletion-image", "rancher/shell:v0.1.6") // need to update the value in chart/values.yaml for any change in this line
```
@cbron this is to address your [comment](https://github.com/rancher/rancher/pull/31670/files#r595497658):
> 1. could work but it needs to load the entire rancher codebase for it to work, so that would take good amount of time.
> 2. If someone forgets then airgap won't work for this, much safer to add to requirements struct and also a comment there. That seems like best solution for now.

cbron replied:
Need to run gofmt on this line. Also I think we should add here and say to update this line: https://github.com/rancher/rancher/blob/master/pkg/image/resolve.go#L248
That resolve.go line needs to always be in sync with values.yaml for this hook, that is our failsafe if someone changes this setting and not the values.yaml.

cbron commented:
I think a single pod with `restartPolicy: OnFailure` is probably best. That will try a bunch of times but only use one pod and if last try fails they still get logs.

jiaqiluo commented:
> I think a single pod with `restartPolicy: OnFailure` is probably best. That will try a bunch of times but only use one pod and if last try fails they still get logs.

@cbron The problem of setting the pod with `restartPolicy: OnFailure` is that the pod will be deleted when the job hits its `backoffLimit` and fails, which gives us no chance to check its logs. 

Quote from https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-backoff-failure-policy

> Note: If your job has restartPolicy = "OnFailure", keep in mind that your container running the Job will be terminated once the job backoff limit has been reached. This can make debugging the Job's executable more difficult. We suggest setting restartPolicy = "Never" when debugging the Job or using a logging system to ensure output from failed Jobs is not lost inadvertently.

jiaqiluo commented:
Based on the conversation with @cbron, we decided to set the pod with `restartPolicy: OnFailure`.

rawmind0 reviewed (dismissed):
LGTM

cbron commented on scripts/chart/build:
```diff
@@ -14,3 +14,9 @@ cp -rf ${1} build/chart/rancher
+
+post_delete_base=$(grep -i shell-image pkg/settings/setting.go | cut -d "," -f 2 | sed -e 's/"//g' | sed -e 's/)//g' | sed -e 's/ //g')
+post_delete_image_name=$(echo ${post_delete_base} | cut -d ":" -f 1)
+post_delete_image_tag=$(echo ${post_delete_base} | cut -d ":" -f 2)
```
Let's add some checks here to ensure the name and tag are what we expect. We _could_ hardcode the image name but thats probably not necessary as it could change down the line. But checking that tag starts with v or that "shell" is in image name, something like that.

jiaqiluo replied:
The following 2 changes are made to address your comments:
1. For the line 18-20, those commands will not exit with code 0, they return an empty string and exit code 0 if fail.
2. The checks for the image name and tag are added at lines 21-28.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,57 @@
+
+succeeded=()
+failed=()
+echo "Uninstalling Rancher resources in the following namespaces:" "$namespaces"
```
```suggestion
echo "Uninstalling Rancher resources in the following namespaces: ${namespaces}"
```

aiyengar2 replied:
Can you change all env var references to `${ENVVAR}` instead of `$ENVVAR`? 

The bracket notation is less ambiguous and imo is generally better practice in bash scripts since you don't encounter issues like `${foo}bar` v.s. `$foobar`.

jiaqiluo replied:
fixed all occurrences

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,57 @@
+
+set -e
+
+succeeded=()
```
Need to add a check to ensure that the expected env vars exists. Example:

```suggestion
if [ -z "${namespaces}" ]; then
  echo "No namespaces provided."
  exit 0
fi

succeeded=()
```

jiaqiluo replied:
fixed.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,57 @@
+failed=()
+echo "Uninstalling Rancher resources in the following namespaces:" "$namespaces"
+for namespace in $namespaces; do
+  app_list=$(helm list -n "$namespace" -q)
+  for app in $app_list; do
```
nit:
```suggestion
  for app in $(helm list -n "${namespace}" -q); do
```

jiaqiluo replied:
fixed.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,57 @@
+      continue
+    fi
+    echo "--- Deleting the app [$app] in the namespace [$namespace]"
+    if [[ ! $(helm uninstall "$app" -n "$namespace") ]]; then
```
Does the following work instead?

```suggestion
    if ! helm uninstall "${app}" -n "${namespace}"; then
```

jiaqiluo replied:
I'd prefer to keep the `[[ ]]` to keep the consistency of the format.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,57 @@
+      continue
+    fi
+    t=0
+    # by default, wait 120 seconds in total for an app to be uninstalled
+    while [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]] && [[ $t -lt "$timeout" ]]; do
+      echo "waiting 5 seconds for pods of [$app] to be terminated ..."
+      sleep 5
+      t=$((t + 5))
+    done
+    if [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]]; then
+      echo "timeout uninstalling [$app] in the namespace [$namespace]"
+      failed=("${failed[@]}" "$app")
+    else
+      echo "successfully uninstalled [$app] in the namespace [$namespace]"
+      succeeded=("${succeeded[@]}" "$app")
+    fi
+  done
```
Can we drop this in favor of just adding the `--timeout` flag on the helm uninstall?

https://helm.sh/docs/helm/helm_uninstall

cbron replied:
We'd have to trap the exit code so we can list all failures. Or do away with that idea.

cbron replied:
Talked with Jack offline and we are going to not use timeout as it doesn't buy us anything but extending default timeout of 5 min.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,57 @@
+    fi
+  done
+  # delete the helm operator pods
+  pod_list=$(kubectl get pods -n "$namespace" -o name --ignore-not-found)
+  for pod in $pod_list; do
```
`--ignore-not-found` is only needed for kubectl delete right? Not for `kubectl get`

```suggestion
  for pod in $(kubectl get pods -n "$namespace" -o name); do
```

jiaqiluo replied:
fixed.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,57 @@
+done
+
+echo "------ Summary ------"
+if [[ ${#succeeded[@]} -ne 0 ]]; then
+  echo "Succeeded to uninstall the following apps:" "${succeeded[@]}"
+fi
+
+if [[ ${#failed[@]} -ne 0 ]]; then
+  echo "Failed to uninstall the following apps:" "${failed[@]}"
+  if [[ "$ignoreTimeoutError" == "false" ]]; then
+    exit 1
+  fi
+else
+  echo "Cleanup finished successfully."
+fi
```
Any reason why we want to collect them and then output it at the end as opposed to just printing an error statement and continuing in the for loop itself? Seems like collecting and dumping at the end makes the code more complicated to maintain (because Bash arrays aren't easy to look at 😅 ).

cbron replied:
I'm good with this either way. It actually does print out as it works or fails. This is just summary.

aiyengar2 commented on chart/templates/post-delete-hook-ServiceAccount.yaml:
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.postDelete.enabled }}
+apiVersion: v1
```
File naming shouldn't be capitalized and have dashes. We should probably just stick to dashes for all of these files for style.

jiaqiluo replied:
yes, I followed the pattern of existing files. But it is a good time to switch to dashes :)

aiyengar2 commented on chart/templates/post-delete-hook-clusterRole.yaml:
```diff
@@ -0,0 +1,28 @@
+    "helm.sh/hook-weight": "1"
+    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded,hook-failed
+rules:
+  - apiGroups: [ "extensions","apps" ]
+    resources: [ "deployments" ]
+    verbs: [ "get", "list", "delete" ]
+  - apiGroups: [ "rbac.authorization.k8s.io" ]
+    resources: [ "clusterroles", "clusterrolebindings" ]
+    verbs: [ "get", "list", "delete" ]
+  - apiGroups: [ "" ]
+    resources: [ "serviceaccounts", "pods", "secrets", "services" ]
+    verbs: [ "get", "list", "delete" ]
+  - apiGroups: [ "networking.k8s.io" ]
+    resources: [ "networkpolicies" ]
+    verbs: [ "get", "list", "delete" ]
+  - apiGroups: [ "admissionregistration.k8s.io" ]
+    resources: [ "validatingwebhookconfigurations" ]
+    verbs: [ "get", "list", "delete" ]
```
nit: to avoid mixing JSON / YAML, I'd personally prefer if this was something like
```yaml
- apiGroups:
  - extensions
  - apps
- resources:
  - deployments
verbs:
  - get
  - list
  - delete
```

If you disagree, feel free to resolve this comment.

jiaqiluo replied:
I followed the format of [examples](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-example) from the K8bernetes documentation, such like the following one:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: default
  name: pod-reader
rules:
- apiGroups: [""] # "" indicates the core API group
  resources: ["pods"]
  verbs: ["get", "watch", "list"]
```

aiyengar2 commented on chart/templates/post-delete-hook-clusterRole.yaml:
```diff
@@ -0,0 +1,28 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ .Values.postDelete.name }}
```
Why does the user need to provide `.postDelete.name`? Let's use something like `{{ template "rancher.fullname" . }}-post-delete` instead if possible and remove the provided name entirely from the values.yaml.

aiyengar2 replied:
same comment for everywhere else this is used.

jiaqiluo replied:
fixed.

aiyengar2 commented on chart/templates/post-delete-hook-clusterRole.yaml:
```diff
@@ -0,0 +1,28 @@
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ .Values.postDelete.name }}
+  labels:
+{{ include "rancher.labels" . | indent 4 }}
```
nit:
```suggestion
  labels: {{ include "rancher.labels" . | nindent 4 }}
```

aiyengar2 replied:
same comment for everywhere else this is used.

jiaqiluo replied:
fixed all occurrences.

aiyengar2 commented on chart/templates/post-delete-hook-configMap.yaml:
```diff
@@ -0,0 +1,16 @@
+data:
+  post-delete-hook.sh: |-
+{{ $.Files.Get "scripts/post-delete-hook.sh" | indent 4 }}
+{{ end }}
```
```suggestion
{{- end }}
```

aiyengar2 replied:
same comment for wherever else the prefix `-` is omitted.

While it's not a big deal, if someone were to add a resource under this it could potentially leave weird newlines in the helm template produced.

jiaqiluo replied:
fixed all occurrences.

aiyengar2 commented on chart/templates/post-delete-hook-job.yaml:
```diff
@@ -0,0 +1,47 @@
+  template:
+    metadata:
+      name: {{ .Values.postDelete.name }}
+      labels:
+        heritage: {{ .Release.Service }}
+        release: {{ .Release.Name }}
```
```suggestion
      labels: {{ include "rancher.labels" . | nindent 8 }}
```

jiaqiluo replied:
fixed.

aiyengar2 commented on chart/templates/post-delete-hook-job.yaml:
```diff
@@ -0,0 +1,47 @@
+          command:
+            - /script/post-delete-hook.sh
+          volumeMounts:
+            - mountPath: /script
```
In case we add more scripts in the future. Will also need to change the command to `/scripts/post-delete-hook.sh`

```suggestion
            - mountPath: /scripts
```

jiaqiluo replied:
fixed.

aiyengar2 commented on chart/templates/post-delete-hook-job.yaml:
```diff
@@ -0,0 +1,47 @@
+            - mountPath: /script
+              name: config-volume
+          env:
+            - name: namespaces
+              value: {{ .Values.postDelete.namespaceList | join " " | quote }}
+            - name: timeout
+              value: {{ .Values.postDelete.timeout | quote }}
+            - name: ignoreTimeoutError
+              value: {{ .Values.postDelete.ignoreTimeoutError | quote }}
```
environment variables that come from outside the bash script should be capitalized, e.g. `NAMESPACES`.

Also, are the `| quote` steps necessary?

jiaqiluo replied:
> environment variables that come from outside the bash script should be capitalized, e.g. NAMESPACES. 

This is fixed. 

> Also, are the | quote steps necessary?

THe `helm uninstall rancher -n cattle-system` will throw the following error if there are no quotes around those env vars

`>  helm uninstall rancher -n cattle-system
Error: uninstallation completed with 1 error(s): warning: Hook post-delete rancher/templates/post-delete-hook-job.yaml failed: Job in version "v1" cannot be handled as a Job: v1.Job.Spec: v1.JobSpec.Template: v1.PodTemplateSpec.Spec: v1.PodSpec.Containers: []v1.Container: v1.Container.v1.Container.Env: []v1.EnvVar: v1.EnvVar.Value: ReadString: expects " or n, but found 1, error found in #10 byte of ...|,"value":120},{"name|..., bigger context ...|cher-operator-system"},{"name":"TIMEOUT","value":120},{"name":"IGNORETIMEOUTERROR","value":false}],"|...
`

aiyengar2 commented on chart/values.yaml:
```diff
@@ -136,3 +136,19 @@ customLogos:
+    repository: %POST_DELETE_IMAGE_NAME%
+    tag: %POST_DELETE_IMAGE_TAG%
+  namespaceList:
+    - fleet-system
+    - cattle-system
+    - rancher-operator-system
```
any reason why we expose this instead of hard-coding the namespaceList in the chart deployment? Is there a use case where a suer can install stuff onto non-default namespaces that we want to support?

cbron replied:
Yes I think Raul requested that. Can also be used to clean up other resources, or not clean up fleet.

aiyengar2 commented on scripts/chart/build:
```diff
@@ -14,3 +14,17 @@ cp -rf ${1} build/chart/rancher
 sed -i -e "s/%VERSION%/${CHART_VERSION}/g" build/chart/rancher/Chart.yaml
 sed -i -e "s/%APP_VERSION%/${VERSION}/g" build/chart/rancher/Chart.yaml
+
+post_delete_base=$(grep -i shell-image pkg/settings/setting.go | cut -d "," -f 2 | sed -e 's/"//g' | sed -e 's/)//g' | sed -e 's/ //g') || ""
```
I see what you are doing here, but I'm really not in favor of running sed through Go code to infer the base image of this chart. IMO, we should be just hard-coding the base image in the chart instead of relying on build magic.

The Version / AppVersion are different since they are small sed commands, but this one is really hard to parse through (because parsing a Go file is not something bash scripts can / should do). This is also relatively fragile since it forces us to never add something with the word `shell-image` before [this](https://github.com/rancher/rancher/blob/b0d3575cc216c5191590b39f23f2888f4e731a69/pkg/settings/setting.go#L97) in our Go code, which is a weird constraint to have on our code.

Was this a strict requirement for this issue?

cbron replied:
No it was just best solution we came up with, grepping for the image. There's no easy way to get to it from this point in build process. However I do like idea of just hardcoding the image name and then parsing for image version. We had thought of that but went this route so that if imaged changed it would catch it, but you are right its very fragile.

jiaqiluo replied:
the PR is updated. Now the image name is hardcoded, and only the image tag is grepped from the go file.

aiyengar2 replied:
If we are anyways grepping the image tag, we're imposing a constraint on the Go code regardless of whether we get the image name or not. So IMO we might as well go back to grepping the image name itself as well; my initial preference was to remove this sed script entirely in favor of hard-coding both the image and image tag.

However, I do have one idea: since rancher/rancher is using [Go 1.16](https://github.com/rancher/rancher/blob/baa461856167e28afea67c12667e92b3f242f467/go.mod#L3) which supports [`go:embed`](https://blog.carlmjohnson.net/post/2021/how-to-use-go-embed/), maybe the solution here is to set

```go
const (
        //go:embed shell_image.txt
	ShellDockerImage string
        //go:embed shell_image_tag.txt
        ShellDockerTag string
)

var (
// ...
	ShellImage                        = NewSetting(ShellDockerImage, ShellDockerTag)
// ...
)
```

That way we can just `cat pkg/settings/shell_image.txt` and `cat pkg/settings/shell_image_tag.txt` to get the image and tag for this script instead of sedding.

Otherwise, let's go back to grepping both the image tag + image name.

Any thoughts @cbron @jiaqiluo?

jiaqiluo replied:
We grep the image name and tag from `pkg/settings/setting.go` because we want to avoid inconsistency and extra dependencies. Using `go:embed` means we will have 2 more files to maintain which is not an ideal solution. 

Going back to grepping both the image tag + image name looks like an acceptable workaround to me.

aiyengar2 replied:
Agreed

cbron replied:
Thats very interesting solution Arvind. We could also just use a const there with a string to make grepping settings.go easier. But I am good with this solution for now.

cbron reviewed (dismissed):
lgtm, once arvind approves we are good to merge and backport to 2.5

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+failed=()
+
+get_pod_count() {
+  kubectl get pods --selector app="${1}" -n "${2}" -o json | jq '.items | length'
```
Can we assume rancher/shell has jq in it?

jiaqiluo replied:
yes, `jq` is included in the rancher/shell image

https://github.com/rancher/shell/blob/49ecd409727764d202026ab7f28b6820c2e28b5e/package/Dockerfile#L15
```
RUN apk add -U --no-cache bash bash-completion jq
```

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+
+for namespace in ${namespaces}; do
+  for app in $(helm list -n "${namespace}" -q); do
+    if [[ ${app} =~ .crd$ ]]; then
```
```suggestion
    if [[ "${app}" =~ ".crd$" ]]; then
```

jiaqiluo replied:
The double brackets `[[` prevents the word splitting by default, so the quotes around the string variable are optional. 
I replaced other `[`s with `[[`s in the file and no quotes to make use of the good feature of  `[[`.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+    fi
+    t=0
+    # by default, wait 120 seconds in total for an app to be uninstalled
+    while [[ $(get_pod_count "${app}" "${namespace}") != "0" ]] && [[ "${t}" -lt "${timeout}" ]]; do
```
```suggestion
    while [[ "$(get_pod_count ${app} ${namespace})" -ne 0 ]] && [[ "${t}" -lt "${timeout}" ]]; do
```

jiaqiluo replied:
1. The double brackets `[[` prevents the word splitting by default, so the quotes around the string variable are optional. 
I replaced other `[`s with `[[`s in the file and no quotes to make use of the good feature of  `[[`. 

2. `"$(get_pod_count ${app} ${namespace})"` returns a string not a number, so we need to compare two strings here.

aiyengar2 replied:
1. Sounds good! Thanks for the clarification
2. Bash doesn’t have the concept of types so strings or numbers don’t make a difference. A number is a “string” with just digits

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+      continue
+    fi
+    echo "--- Deleting the app [${app}] in the namespace [${namespace}]"
+    if [[ ! $(helm uninstall "${app}" -n "${namespace}") ]]; then
```
Check if this works:

```suggestion
    if ! helm uninstall "${app}" -n "${namespace}"; then
```

jiaqiluo replied:
it works with or without the double brackets, but I'd prefer to keep the brackets to keep the style consistent.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+      sleep 5
+      t=$((t + 5))
+    done
+    if [[ $(get_pod_count "${app}" "${namespace}") != "0" ]]; then
```
```suggestion
    if [[ "$(get_pod_count ${app} ${namespace})" -ne 0 ]]; then
```

aiyengar2 replied:
If the deletion is successful the first time around, this seems like an unnecessary second call to be made. Could we move the failure logic directly into the previous while statement?

e.g. 

```shell
while : ; do
    pod_count=$(get_pod_count ${app} ${namespace})
    if [[ "${pod_count}" -eq 0 ]]; then
        echo "successfully uninstalled [${app}] in the namespace [${namespace}]"
        succeeded=("${succeeded[@]}" "${app}")
        break
    fi
    if [[ "${t}" -ge "${timeout}" ]]; then
        echo "timeout uninstalling [${app}] in the namespace [${namespace}]"
        failed=("${failed[@]}" "${app}")
        break
    fi
    echo "waiting 5 seconds for pods of [${app}] to be terminated ..."
    sleep 5
    t=$((t + 5))
done
```

(note: `while : ` is bash for `while true`)

jiaqiluo replied:
Good point! The script is updated to use this `while` loop.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+  done
+  # delete the helm operator pods
+  for pod in $(kubectl get pods -n "${namespace}" -o name); do
+    if [[ ${pod} =~ ^pod\/helm-operation-* ]]; then
```
```suggestion
    if [[ "${pod}" =~ ^pod\/helm-operation-* ]]; then
```

jiaqiluo replied:
The double brackets `[[` prevents the word splitting by default, so the quotes around the string variable are optional. 
I replaced other `[`s with `[[`s in the file and no quotes to make use of the good feature of  `[[`.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+done
+
+echo "------ Summary ------"
+if [[ ${#succeeded[@]} -ne 0 ]]; then
```
```suggestion
if [[ "${#succeeded[@]}" -ne 0 ]]; then
```

jiaqiluo replied:
The double brackets `[[` prevents the word splitting by default, so the quotes around the string variable are optional. 
I replaced other `[`s with `[[`s in the file and no quotes to make use of the good feature of  `[[`.

aiyengar2 commented on chart/scripts/post-delete-hook.sh:
```diff
@@ -0,0 +1,80 @@
+  echo "Succeeded to uninstall the following apps:" "${succeeded[@]}"
+fi
+
+if [[ ${#failed[@]} -ne 0 ]]; then
```
```suggestion
if [[ "${#failed[@]}" -ne 0 ]]; then
```

jiaqiluo replied:
The double brackets `[[` prevents the word splitting by default, so the quotes around the string variable are optional. 
I replaced other `[`s with `[[`s in the file and no quotes to make use of the good feature of  `[[`.

aiyengar2 commented on scripts/chart/build:
```diff
@@ -14,3 +14,11 @@ cp -rf ${1} build/chart/rancher
+
+post_delete_base=$(grep -i shell-image pkg/settings/setting.go | cut -d "," -f 2 | sed -e 's/"//g' | sed -e 's/)//g' | sed -e 's/ //g') || ""
+post_delete_image_tag=$(echo "${post_delete_base}" | cut -d ":" -f 2) || ""
+if [[ ! $post_delete_image_tag =~ ^v.* ]]; then
```
```suggestion
if [[ ! ${post_delete_image_tag} =~ ^v.* ]]; then
```

aiyengar2 commented on scripts/chart/build:
```diff
@@ -14,3 +14,11 @@ cp -rf ${1} build/chart/rancher
 sed -i -e "s/%VERSION%/${CHART_VERSION}/g" build/chart/rancher/Chart.yaml
 sed -i -e "s/%APP_VERSION%/${VERSION}/g" build/chart/rancher/Chart.yaml
+
+post_delete_base=$(grep -i shell-image pkg/settings/setting.go | cut -d "," -f 2 | sed -e 's/"//g' | sed -e 's/)//g' | sed -e 's/ //g') || ""
```
nit: add a comment on top of this gigantic sed command for humans to read 😄

jiaqiluo replied:
A comment is added

aiyengar2 reviewed (approved):
I have two nit comments and @jiaqiluo needs to add back in the image stuff in https://github.com/rancher/rancher/pull/31670/files#r603693702, but LGTM.

Great work!! 🎉 🎉
